### PR TITLE
feat: more precise pegin verifier status & state machine

### PR DESCRIPTION
**Pull request by sander2**
For the peg-ins, all verifiers need to sign off on the PegInConfirm. This PR adds more states to the `PegInVerifierStatus` and adds appropriate actions for them. General flow for a given verifier is:

- If we have not sent our nonces, send them
- wait until all verifiers have sent nonces
- if we have not sent our signature, send it
- wait until all verifiers have sent signatures
- submit the `PegInConfirm`

There is a test that demonstrates the process. Assuming this PR gets merged I will make a follow-up pr to add the peg-in handling to the cli.

### CODE REVIEW

File: src/bridge/client/cli/client_command.rs [hunk 1]
```diff
@@ -73,32 +73,8 @@ impl ClientCommand {
         loop {
             self.client.sync().await;
 
-            let peg_in_graphs = self.client.get_data().peg_in_graphs.clone();
 
-            for peg_in_graph in peg_in_graphs.iter() {
```
Comment (StarDumpling) [inline src/bridge/client/cli/client_command.rs:77]: `handle_automatic_command` fn should run all actions depending on client's role (one client can have more than one role). 

We're currently missing handling verifier's commands. However we should keep depositor's operations here as we'll complete the function with verifier's commands soon.

File: src/bridge/graphs/peg_in.rs [hunk 2]
```diff
@@ -58,20 +58,35 @@ impl Display for PegInDepositorStatus {
...
+    Complete,            // peg-in complete
 }
 
 impl Display for PegInVerifierStatus {
     fn fmt(&self, f: &mut Formatter) -> FmtResult {
         match self {
-            PegInVerifierStatus::PegInWait => write!(f, "No action available. Wait..."),
-            PegInVerifierStatus::PegInPresign => {
-                write!(f, "Signature required. Presign peg-in confirm transaction?")
+            PegInVerifierStatus::AwaitingDeposit => write!(f, "Waiting for deposit transaction..."),
```
Comment (StarDumpling) [inline src/bridge/graphs/peg_in.rs]: To be more consistent with the rest of the code I suggest:
```suggestion
            PegInVerifierStatus::AwaitingDeposit => write!(f, "Peg-in deposit transaction not confirmed yet. Wait..."),
```

File: src/bridge/graphs/peg_in.rs [hunk 3]
```diff
@@ -58,20 +58,35 @@ impl Display for PegInDepositorStatus {
...
+            PegInVerifierStatus::ReadyToSubmit => {
+                write!(f, "Peg-in pre-signed, confirm transaction ready.")
```
Comment (StarDumpling) [inline src/bridge/graphs/peg_in.rs]: All the statuses should either tell the user that something is done, they should wait as there is no action available at the moment or there is an action available and ask if they want to proceed.

```suggestion
                write!(f, "Peg-in confirm transaction pre-signed. Broadcast confirm transaction?")
```

File: src/bridge/graphs/peg_in.rs [hunk 4]
```diff
@@ -58,20 +58,35 @@ impl Display for PegInDepositorStatus {
...
             }
-            PegInVerifierStatus::PegInComplete => write!(f, "Peg-in complete. Done."),
+            PegInVerifierStatus::PendingOurNonces => write!(f, "Peg-in pending our nonces."),
```
Comment (StarDumpling) [inline src/bridge/graphs/peg_in.rs]: I suggest:
```suggestion
            PegInVerifierStatus::PendingOurNonces => write!(f, "Nonce required. Share peg-in confirm nonce?"),
```

File: src/bridge/graphs/peg_in.rs [hunk 5]
```diff
@@ -58,20 +58,35 @@ impl Display for PegInDepositorStatus {
...
+            PegInVerifierStatus::AwaitingNonces => {
+                write!(f, "Peg-in pending verifier nonces.")
+            }
+            PegInVerifierStatus::PendingOurSignature => {
+                write!(f, "Peg-in pending our signature.")
```
Comment (StarDumpling) [inline src/bridge/graphs/peg_in.rs]: To be more consistent with the codebase, I suggest:
```suggestion
                write!(f, "Signature required. Pre-sign peg-in confirm transaction?")
```

File: src/bridge/graphs/peg_in.rs [hunk 6]
```diff
@@ -58,20 +58,35 @@ impl Display for PegInDepositorStatus {
...
-            PegInVerifierStatus::PegInPresign => {
-                write!(f, "Signature required. Presign peg-in confirm transaction?")
+            PegInVerifierStatus::AwaitingDeposit => write!(f, "Waiting for deposit transaction..."),
+            PegInVerifierStatus::ReadyToSubmit => {
+                write!(f, "Peg-in pre-signed, confirm transaction ready.")
             }
-            PegInVerifierStatus::PegInComplete => write!(f, "Peg-in complete. Done."),
+            PegInVerifierStatus::PendingOurNonces => write!(f, "Peg-in pending our nonces."),
+            PegInVerifierStatus::AwaitingNonces => {
+                write!(f, "Peg-in pending verifier nonces.")
```
Comment (StarDumpling) [inline src/bridge/graphs/peg_in.rs]:
```suggestion
                write!(f, "Awaiting peg-in confirm nonces. Wait...")
```

File: src/bridge/graphs/peg_in.rs [hunk 7]
```diff
@@ -58,20 +58,35 @@ impl Display for PegInDepositorStatus {
...
+            }
+            PegInVerifierStatus::PendingOurSignature => {
+                write!(f, "Peg-in pending our signature.")
+            }
+            PegInVerifierStatus::AwaitingSignatures => {
+                write!(f, "Peg-in pending verifier signatures.")
```
Comment (StarDumpling) [inline src/bridge/graphs/peg_in.rs]:
```suggestion
                write!(f, "Awaiting peg-in confirm signatures. Wait...")
```

File: src/bridge/transactions/peg_in_confirm.rs [hunk 8]
```diff
@@ -167,6 +187,8 @@ impl PegInConfirmTransaction {
         let secret_nonce = push_nonce(self, context, input_index);
         secret_nonces.insert(input_index, secret_nonce);
 
+        println!("Pusing nonce for {}", context.verifier_public_key);
```
Comment (StarDumpling) [inline src/bridge/transactions/peg_in_confirm.rs]: To be more consistent with the codebase, I would remove this `println`.

File: src/bridge/client/client.rs [hunk 9]
```diff
@@ -612,6 +612,65 @@ impl BitVMClient {
...
+                    );
+                }
+            }
+        }
+    }
+
+    pub async fn process_peg_in_as_verifier(&mut self, peg_in_graph: &PegInGraph) {
+        if let Some(ref context) = self.verifier_context {
+            let status = peg_in_graph.verifier_status(&self.esplora, Some(&context)).await;
+            println!("Status: {}", status);
```
Comment (StarDumpling) [inline src/bridge/client/client.rs]: This `println` should be removed.

File: src/bridge/client/client.rs [hunk 10]
```diff
@@ -612,6 +612,65 @@ impl BitVMClient {
         }
     }
 
+    pub async fn process_peg_in_as_depositor(&mut self,peg_in_graph: &PegInGraph) {
```
Comment (StarDumpling) [inline src/bridge/client/client.rs:614]: This function and `process_peg_in_as_verifier()` function belong to client_command.rs file as we handle user's commands there. Or we could create such functions for all client roles in both peg-in and peg-out graphs and use them in `handle_automatic_command()` function.

**Review by StarDumpling — CHANGES_REQUESTED** [review]

File: src/bridge/client/cli/client_command.rs (hunk 1 above)
Comment (justin-elementlabs) [reply src/bridge/client/cli/client_command.rs:77]: Right, this is a little bit of a confusing point. Our client:

• supports as roles: depositor, operator, verifier, withdrawer
• functions are implemented piece-wise per role; and a client can startup the client with multiple roles running concurrently

Comment (sander2) [reply src/bridge/client/cli/client_command.rs:77]: Woops, my bad. I forgot to call `process_peg_ins` here - fixed now.

I also extracted the pegout processing logic into a `process_peg_outs` function which I placed into the client file as well. I think it makes more sense to have most of the logic in the client file, and have `client_command` just be an entry point into the program. If you disagree just lmk and I'll refactor it back

File: src/bridge/client/client.rs (hunk 10 above)
Comment (sander2) [reply src/bridge/client/client.rs:614]: Please see the comment above

File: src/bridge/client/cli/client_command.rs (hunk 1 above)
Comment (StarDumpling) [reply src/bridge/client/cli/client_command.rs:77]: I'm fine with this refactor.

**Pull request merged**